// @angular-devkit/schematics/src/engine/schematic.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const exception_1 = require("../exception/exception");
const Observable_1 = require("rxjs/Observable");
require("rxjs/add/observable/of");
require("rxjs/add/operator/concatMap");
class InvalidSchematicsNameException extends exception_1.BaseException {
    constructor(name) {
        super(`Schematics has invalid name: "${name}".`);
    }
}
exports.InvalidSchematicsNameException = InvalidSchematicsNameException;
class SchematicImpl {
    constructor(_description, _factory, _collection, _engine) {
        this._description = _description;
        this._factory = _factory;
        this._collection = _collection;
        this._engine = _engine;
        if (!_description.name.match(/^[-_.a-zA-Z0-9]+$/)) {
            throw new InvalidSchematicsNameException(_description.name);
        }
    }
    get description() { return this._description; }
    get collection() { return this._collection; }
    call(options, host) {
        let context = {
            engine: this._engine,
            schematic: this,
            strategy: this._engine.defaultMergeStrategy
        };
        const transformedOptions = this._engine.transformOptions(this, options);
        return host.concatMap(tree => {
            const result = this._factory(transformedOptions)(tree, context);
            if (result instanceof Observable_1.Observable) {
                return result;
            }
            else {
                return Observable_1.Observable.of(result);
            }
        });
    }
}
exports.SchematicImpl = SchematicImpl;
